nomad/resource_job_v2.go

```diff
@@ -0,0 +1,1682 @@
+			"value":     cn.RTarget,
+		}
+
+		res = append(res, constraint)
+	}
+
+	return res
+}
+
+func readAffinities(affinities []*api.Affinity) interface{} {
+	res := make([]interface{}, 0)
+
+	for _, af := range affinities {
+		affinity := map[string]interface{}{
+			"attribute": af.LTarget,
+			"operator":  af.Operand,
+			"value":     af.RTarget,
+			"weight":    af.Weight,
+		}
+
+		res = append(res, affinity)
+	}
+
+	return res
+}
+
+func readSpreads(spreads []*api.Spread) interface{} {
+	res := make([]interface{}, 0)
+
+	for _, s := range spreads {
+		targets := make([]interface{}, 0)
+
+		for _, t := range s.SpreadTarget {
+			target := map[string]interface{}{
+				"value":   t.Value,
+				"percent": t.Percent,
+			}
+
+			targets = append(targets, target)
+		}
+
+		spread := map[string]interface{}{
+			"attribute": s.Attribute,
+			"weight":    s.Weight,
+			"target":    targets,
+		}
+
+		res = append(res, spread)
+	}
+
+	return res
+}
+
+func readGroups(d *schema.ResourceData, groups []*api.TaskGroup) (interface{}, error) {
+	res := make([]interface{}, 0)
+
+	// we have to look for the groups the user created in its configuration as
+	// we will need to set the "ephemeral_disk" and "restart" block only if he
+	// created one or the value for the block is different from the default one
+
+	groupsConfig := d.Get("group").([]interface{})
+
+	for i, g := range groups {
+		currentConfig := groupsConfig[i].(map[string]interface{})
+
+		ephemeralDisk := make([]interface{}, 0)
+
+		if g.EphemeralDisk != nil {
+			disk := map[string]interface{}{
+				"migrate": *g.EphemeralDisk.Migrate,
+				"size":    *g.EphemeralDisk.SizeMB,
+				"sticky":  *g.EphemeralDisk.Sticky,
+			}
+
+			defaultValue := map[string]interface{}{
+				"migrate": false,
+				"size":    300,
+				"sticky":  false,
+			}
+
+			set := currentConfig["ephemeral_disk"].([]interface{})
+
+			if len(set) > 0 || !reflect.DeepEqual(disk, defaultValue) {
+				ephemeralDisk = append(ephemeralDisk, disk)
+			}
+		}
+
+		restart := make([]interface{}, 0)
+		if g.RestartPolicy != nil {
+			r := map[string]interface{}{
+				"attempts": *g.RestartPolicy.Attempts,
+				"delay":    g.RestartPolicy.Delay.String(),
+				"interval": g.RestartPolicy.Interval.String(),
+				"mode":     *g.RestartPolicy.Mode,
+			}
+
+			// The default depends on the job type
+			var defaultValue map[string]interface{}
+			_type := d.Get("type").(string)
+			if _type == "service" {
+				defaultValue = map[string]interface{}{
+					"attempts": 2,
+					"delay":    "15s",
+					"interval": "30m0s",
+					"mode":     "fail",
+				}
+			} else if _type == "batch" {
+				defaultValue = map[string]interface{}{
+					"attempts": 3,
+					"delay":    "15s",
+					"interval": "24h0m0s",
+					"mode":     "fail",
+				}
+			} else if _type == "system" {
+				defaultValue = map[string]interface{}{
+					"attempts": 2,
+					"delay":    "15s",
+					"interval": "30m0s",
+					"mode":     "fail",
+				}
```

Comment:
Does Nomad require these values to be set? If possible, it would be better to submit the job without them and let Nomad set their defaults.

Reply:
This is actually linked to https://github.com/hashicorp/terraform-plugin-sdk/issues/62. If the user did not set a `restart policy it is automatically set by Nomad and returned in the job spec. In this case Terraform gets a list of one element but does not expect any a perpetually creates a diff for this resource.

We had this issue in the Consul provider (https://github.com/hashicorp/terraform-provider-consul/pull/121 and https://github.com/hashicorp/terraform-provider-consul/issues/119) and the only solution I could find was to check whether the user had set the attribute and if the server had sent back the default values for this attribute and ignore it in this case. It's cumbersome but I'm not aware of a better solution at the moment. Does this make sense?